FxA OAuth flow

sashei opened the pull request:
resolves #31

sashei commented:
not ready for review or merge yet

rfk commented on lockbox-ios/FxA/FxAPresenter.swift:
```diff
@@ -0,0 +1,178 @@
+    private var scopedKeySubject = PublishSubject<OAuthInfo>()
+
+    internal let oauthHost = "oauth-scoped-keys-oct10.dev.lcip.org"
+    internal let redirectURI = "lockbox://redirect.ios"
```
Sorry for the drive-by, but I noticed this redirectURI wile reviewing the FxA-side code and I wanted to flag it for follow-up.

One of the issues that came up in the external security review of the scoped-keys work, was that there's a measurable difference in security between using URIs with a custom protocol (like lockbox://redirect.ios) and using https URIs delegated to the native app via universal links (like e.g. https://lockbox.firefox.com/oauth/ios coupled with [1]).  We should aim to use https URIs for redirects wherever possible, because it makes it much harder for a malicious iOS app to try to intercept redirects intended for the lockbox app.

Something to consider for a post-beta follow-up perhaps; I just wanted to put it on your radar while I thought of it :-)

[1] https://developer.apple.com/library/content/documentation/General/Conceptual/AppSearch/UniversalLinks.html

sashei replied:
Nice catch, Ryan. I'm not too worried about the redirectURI for alpha because in the sandboxed WKWebView that I'm using in this implementation, there's no system-level call for any web redirects. We're also using PKCE specifically to protect against the type of malicious redirect interception you mention.

Changing the redirectURI we're using in the sandboxed webview is trivial, though, and would just require a config change in the scoped-keys development environment (@vladikoff might be able to help us out there). If we want to guard against (or support ;) a system-level universal link redirect for something like https://lockbox.firefox.com, there's more work involved but you're right that it's worth looking into for beta.

vladikoff replied:
Yeah just let me know if redirect uri needs to be adjusted

linuxwolf commented on lockbox-ios/Models/KeychainManager.swift:
```diff
@@ -0,0 +1,46 @@
+    case email, scopedKey
+}
+
+class KeychainManager {
```
it is important to also save the FxA uid; it is used by `lockbox-datastore` as the salt for deriving the encryption key and hashing key

sashei commented:
addressed comments; currently, the datastore initialization & hookup will be dependent on other architecture decisions so using the uid & scoped key is not implemented yet (tracked in issue #87)
